### ISSUE TITLE
plot -Z broken in Julia

### ISSUE THREAD

**Issue by joa-quim** [GenericMappingTools/gmt#6133]
This example no longer works.
https://www.generic-mapping-tools.org/GMT.jl/dev/gallery/choropleths/choropleth_cv19/

The problem may be due to changes in KEYs parsing as this chunk in ``psxy.c``
```
case 'Z':		/* Get value for CPT lookup */
	n_errors += gmt_M_repeated_module_option (API, Ctrl->Z.active);
	Ctrl->Z.active = true;
	if (gmt_not_numeric (GMT, opt->arg) && !gmt_access (GMT, opt->arg, R_OK)) {	/* Got a file */
		Ctrl->Z.file = strdup (opt->arg);
		n_errors += gmt_M_check_condition (GMT, Ctrl->Z.file && gmt_access (GMT, Ctrl->Z.file, R_OK),
		                                  "Option -Z: Cannot read file %s!\n", Ctrl->Z.file);
	}
```
has ``opt->arg = ""`` and the ``Ctrl->Z.file`` is not set so no Z is transmitted.

**Comment by joa-quim** [comment]
Confirmed because 

this fails
```
plot(D, R="-1/3/-1/3", J="x2c", B=0, Z=[1,4], G="+z", C=C)
```
but saving the ``[1,4]`` in a file then this works
```
plot(D, R="-1/3/-1/3", J="x2c", B=0, Z="z.dat", G="+z", C=C)
```

**Comment by PaulWessel** [comment]
Trying to make sense of this section (which is where the KEY `ZD(= `is going.  The _gmtapi_extract_argument_ call does that.  It finds the = with nothing after it and all it does is to set implicit, meaning we expect an implicit input file.

The gmt_access returns 0 if given a memory file so the Z.file should be strdup'ed here.  What are you passing in the command?

**Comment by PaulWessel** [comment]
The key is correct and no need for anyone to have changed that since you are the only one with test cases for this.  Are you setting and passing a memory file via -Z? If so then opt->arg cannot be not set (GMT does not remove things).

**Comment by joa-quim** [comment]
> What are you passing in the command?

I showed above, ``Z=[1,4]``

If you click the example you see that it has worked before.

**Comment by PaulWessel** [comment]
yes, but you are not passing [1,4] to Call_Module.  You are creating a virtual data set and passing its name, no?

**Comment by PaulWessel** [comment]
I am asking since I do not understand if you passed -Z@GMTAPI-########### why opt->arg can be empty.

**Comment by joa-quim** [comment]
I'm doing the exact the same thing as with all the other options that pass memory data. The functions are generic, no special treatment for this -Z option.

**Comment by joa-quim** [comment]
See, ``opt->arg`` is empty.

![image](https://user-images.githubusercontent.com/537321/146704735-062ee389-cd6f-4597-a568-f9e25e8f6867.png)

**Comment by PaulWessel** [comment]
OK, I guess I need to fire up Julia and see if I can still do Xcode (macOS and Xcode update since last time).  My suspicion has to do with some checks for remote global grids which start with `@` and then we have the `@GMTAPI` start which perhaps I forgot to exempt...

**Comment by PaulWessel** [comment]
If you are already in the debug, follow it into the gmt_init_module and see where the opt->arg gets eaten.

**Comment by joa-quim** [comment]
Can't debug Julia. Takes an eternity to step.

**Comment by PaulWessel** [comment]
OK, but you are able set a verbose flag to print out the cmd string for Call_Module, no?

**Comment by joa-quim** [comment]
But there is no problem there. Data is well transmitted by the generic code. You keep doubting this.

Sometimes the program crash and following one of those crashes I found that the Z values are in fact inside the lib. The psxy.c line 2366
```
for (tbl = 0; tbl < D->n_tables; tbl++) {
```
has ``n_tables`` = 2 and the second table holds the column vector with the Z values. Then latter down and because ``polygon`` = true it crashes when it calls ``gmtplot_at_pole()`` because there is no ``*lat``.

So the Z data was indeed read but landed on the wrong place.

![Capture](https://user-images.githubusercontent.com/537321/146773547-9dba5081-b749-454b-ab5b-c703cb888e0b.PNG)

One-liner that shows the problem
```
plot(mat2ds(cat([0. 0; 0 1; 1 1; 1 0.3; 0 0], [1 1; 1 2; 2 2; 2 1; 1 1], dims=3)), R="-1/3/-1/3", J="x2c", B=0, Z=[1.2 0;4 0], G="+z", C=makecpt(C=:hot, T="0/5"))
```

**Comment by joa-quim** [comment]
And if I change the KEY from ``ZD(=`` to ``ZD(`` then it works fine again.

**Comment by PaulWessel** [comment]
Since -Z has no modifiers I wonder if the = is a relic of a previous version of -Z.  I cannot see why it would need anything special either.

**Comment by PaulWessel** [comment]
Please make a PR for the change to the KEY.  I need to add some comments somewhere to explain this black magic.

**Comment by joa-quim** [comment]
That KEY change was part of a larger commit (3/7/2020) that changed other likewise KEYs

* Correct keys settings for files to be read via options (#2857)

* Correct keys settings for files to be read via options

**Comment by PaulWessel** [comment]
Maybe be I need to document this better first before we change.  It is not
at all clear to me now what the = and + and trailing integers really do.
It was clear when I coded it but that braincell died.

On December 20, 2021 at 8:36:20 AM, Joaquim ***@***.***)
wrote:

That KEY change was part of a larger commit (3/7/2020) that changed other
likewise KEYs

* Correct keys settings for files to be read via options (#2857)

* Correct keys settings for files to be read via options

**Comment by Esteban82** [comment]
> @GMTAPI-

Just an off-topic question. Are these remote files that are download?

**Comment by joa-quim** [comment]
No. Those are internal names of the virtual files.

**Comment by Esteban82** [comment]
> No. Those are internal names of the virtual files.

Mm ok. I remember that once I have an error. I was on a PC without internet and I got an error that said I couldn't get that filed. It worked when I connect the PC to internet. I must said that I am not complete sure. How can I test this?

**Comment by PaulWessel** [comment]
Not sure if you can - but should be completely unrelated.  The @GMTAPI-##### are fake filenames that GMT uses to pass memory objects between modules.  It has nothing to do with the remote file mechanism.  For example, when you use grdimage -I+d to auto-generate a shading grid from the input topo, grdimage will set up a call to grdgradient to do that work and it returns a grid via these virtual file names.

**Comment by seisman** [comment]
> > No. Those are internal names of the virtual files.
> 
> Mm ok. I remember that once I have an error. I was on a PC without internet and I got an error that said I couldn't get that filed. It worked when I connect the PC to internet. I must said that I am not complete sure. How can I test this?

There was a bug that internal names `@GMTAPI-xxx` were incorrectly considered as remote files. The bug affected GMT 6.1.0 (and maybe also GMT 6.0.0) but was fixed in GMT >=6.1.1 by PR https://github.com/GenericMappingTools/gmt/pull/3745.

**Comment by meghanrjones** [comment]
Closing based on https://github.com/GenericMappingTools/gmt/pull/6134
